backend/appender/rolling/appender.go

```diff
@@ -0,0 +1,210 @@
+// current active log file.
+type FileStater interface {
+	FileStat() FileInfo
+}
+
+type FileInfo struct {
+	Name    string
+	Size    uint64
+	Created time.Time
+}
+
+func NewAppender(
+	lvl backend.Level,
+	layout layout.Factory,
+	triggerFactory triggerFactory,
+	strategyFactory strategyFactory,
+) (*Appender, error) {
+	a := &Appender{
+		lvl: lvl,
+		background: &Background{
+			done: make(chan struct{}),
+		},
+	}
+
+	l, err := layout(a)
+	if err != nil {
+		return nil, err
+	}
+
+	a.layout = l
+	a.strategy = strategyFactory(a.background, a)
+
+	// trigger factory should be initialized last. All state must be initialized here,
+	// as triggers are allowed to trigger a rotate right on startup.
+	// This will lead to two rotate calls. The first one will try to open the file,
+	// the second one (executed within this constructor) will do the rotation.
+	// If the log file is empty, no rotation will occur on the second rotation signal.
+	a.trigger = triggerFactory(a.background, a, a)
+
+	err = a.Rotate()
+	return a, err
+}
+
+func (a *Appender) Close() error {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	if a.closed {
+		return nil
+	}
+
+	a.background.shutdown()
+
+	a.wg.Wait() // wait for async rotation jobs to finish
+	a.closed = true
+
+	a.background.wait()
+
+	return nil
+}
+
+func (a *Appender) For(name string) backend.Backend {
+	return a
+}
+
+func (a *Appender) IsEnabled(lvl backend.Level) bool {
+	return lvl >= a.lvl
+}
+
+func (a *Appender) UseContext() bool {
+	return a.layout.UseContext()
+}
+
+func (a *Appender) Log(msg backend.Message) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	if a.closed {
+		return
+	}
+
+	a.layout.Log(msg)
+	if a.trigger != nil && a.trigger.CheckTrigger(msg, a.stat) {
+		a.execRotate()
+	}
+}
+
+func (a *Appender) Rotate() error {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.execRotate()
+}
+
+func (a *Appender) execRotate() error {
+	if a.closed {
+		return ErrClosed
+	}
+
+	if a.file != nil {
+		if a.stat.Size == 0 {
+			return nil // file was just rotated -> no action
+		}
+
+		a.file.Close() // TODO: report error in multi-error
+		a.file = nil
+	}
+
+	stat := a.stat
+	a.stat = FileInfo{}
+
+	sync, async := a.strategy.Rotate(stat)
+	file, err := sync(stat)
+	if err != nil {
+		return err
+	}
+
+	fi, err := file.Stat()
+	if err != nil {
+		file.Close()
+		return err
+	}
+
+	sz := uint64(fi.Size())
+	timestamp := fi.ModTime()
+	if fi.Size() <= 0 {
+		sz = 0
+		timestamp = time.Now()
+	}
+
+	a.file = file
+	a.stat = FileInfo{
+		Name:    a.file.Name(),
+		Created: timestamp,
+		Size:    sz,
+	}
+	if async == nil {
+		return nil
+	}
+
+	a.wg.Wait() // wait for rotation from last run to finish
+	a.wg.Add(1)
+	go func() {
+		defer a.wg.Done()
+		async(a, stat) // TODO: collect and report error
+	}()
+
+	return nil
+}
+
+func (a *Appender) FileStat() FileInfo {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.stat
+}
+
+func (a *Appender) Write(b []byte) (int, error) {
+	if a.file == nil {
```

Comment:
exported method, is it safe to omit lock here, this possibly calls rotate?

Reply:
The writer is not called directly from the process, but it is called from the 'layout'. That is the lock is already taken. Will try to make this method 'unexported'.

Reply:
Did update the code such that Write is not exported anymore (plus added some comments).